blivet/devicefactory.py

```diff
@@ -294,50 +301,70 @@ def __init__(self, storage, size, disks, fstype=None, mountpoint=None,
             :keyword container_size: requested container size
-            :type container_size: :class:`~.size.Size`
+            :type container_size: :class:`~.size.Size`, :const:`SIZE_POLICY_AUTO` (the default),
+                                  or :const:`SIZE_POLICY_MAX`
             :keyword min_luks_entropy: minimum entropy in bits required for
                                        LUKS format creation
             :type min_luks_entropy: int
 
         """
-        self.storage = storage          # a Blivet instance
-        self.size = size                # the requested size for this device
-        self.disks = disks              # the set of disks to allocate from
-
-        self.original_size = size
-        self.original_disks = disks[:]
-
-        self.fstype = fstype
-        self.mountpoint = mountpoint
-        self.label = label
-
-        self.raid_level = raid_level
-        self.container_raid_level = container_raid_level
-
-        self.encrypted = encrypted
-        self.container_encrypted = container_encrypted
-
-        self.container_name = container_name
-        self.device_name = name
-
-        self.container_size = container_size
-
-        self.container = None
-        self.device = device
+        self.storage = kwargs["storage"]  # a Blivet instance
+
+        self._encrypted = None
+        self._raid_level = None
+        self._container_raid_level = None
+
+        for setting, value in self._default_settings.items():
+            setattr(self, setting, value)
+
+        self.fstype = None  # not included in default_settings b/c of special handling below
+        self.min_luks_entropy = luks_data.min_entropy
+
+        self.device = kwargs.get("device")
+
+        # Now override defaults with user-specified settings.
+        if "size" in kwargs:
+            self.size = kwargs["size"]
+        if "disks" in kwargs:
+            self.disks = kwargs["disks"]
+
+        if "fstype" in kwargs:
+            self.fstype = kwargs["fstype"]
+        if "mountpoint" in kwargs:
+            self.mountpoint = kwargs["mountpoint"]
+        if "label" in kwargs:
+            self.label = kwargs["label"]
+
+        if "name" in kwargs:
+            self.device_name = kwargs["name"]
+        if "raid_level" in kwargs:
+            self.raid_level = kwargs["raid_level"]
+        if "encrypted" in kwargs:
+            self.encrypted = kwargs["encrypted"]
+
+        if "container_name" in kwargs:
+            self.container_name = kwargs["container_name"]
+        if "container_size" in kwargs:
+            self.container_size = kwargs["container_size"]
+        if "container_raid_level" in kwargs:
+            self.container_raid_level = kwargs["container_raid_level"]
+        if "container_encrypted" in kwargs:
+            self.container_encrypted = kwargs["container_encrypted"]
+
+        if "min_luks_entropy" in kwargs:
+            self.min_luks_entropy = kwargs["min_luks_entropy"]
```

Comment:
Maybe this block could be simplified by creating a copy of the defaults and then calling ``defaults_copy.update(kwargs)`` followed by the for cycle with ``setattr`` above moved here?